### DISCUSSION TITLE
Template with multiple solutions

### DISCUSSION THREAD

**Discussion by neman** [dotnet/templating#5342] [Template authoring] [answered]
I"ve used templating long time ago and created template with multiple projects.

Now it is time to update that template to contain multiple solutions with multiple projects.

For example, I want to have template to be able to create either clean architecture structure or classic layered architecture. Is it possible to do that now?

CLI command should be something like:
`dotnet new MyTemplate --architecture Clean`
`dotnet new MyTemplate --architecture Layered `
`dotnet new MyTemplate --architecture VerticalSlice`

There should be one nuget which supports all this templates.

Or do you recommends that I should have n templates for n architecture styles?

`dotnet new MyTemplate.cClean`
`dotnet new MyTemplate.Layered `
`dotnet new MyTemplate.Vertical.Slice`

**Comment by vlada-shubina** [comment]
Imo, this should be a single template with `architecture` parameter (symbol). In `dotnet new`, the parameter symbols exposed via command line options (e.g. `dotnet new my-template --architecture my-arch`).
To learn on how to define the symbols, please check https://github.com/dotnet/templating/wiki/Reference-for-template.json#parameter-symbol. Since the possible values are limited, I would recommend to use `choice` type (equivalent to `enum`).

**Reply by neman** [reply]
So it is possible for template to contains multiple solutions with multiple projects?

**Reply by vlada-shubina** [reply]
The template package can contain N individual templates, but the templates are still run individually. 
Case when single template contains multiple solutions should work in `dotnet new`, however likely it won't work in Visual Studio as it has additional processing.
@phenning could you please advise? Thank you.

**Comment by phenning** [comment, marked as answer]
Visual Studio will handle multiple slns being output with the following cavets:

1. You specify the solution file you want loaded by Visual Studio as a primary output. If you do not specify a primary output and multiple slns are detected, we will just add all generated proj files to the Visual Studio solution. If only a single sln is found, VS will use that one.
2. The solution file should be named so as to not conflict with the Visual Studio generated solution: https://github.com/sayedihashimi/template-sample#solution-name-best-practices
3. Realize that the solution file mentioned above will be *merged* with the Visual Studio created solution.
4. After the solution file is merged, the original sln from the template is deleted.
5. The other slns generated will not be visible by default in solution explorer, unless file view is selected.

Given that, my suggestion would be to either have one template with choice parameters like Vlada suggested to control which solutions to generate. These can be exposed both on the command line and in Visual Studio. You can use the conditions to use source excludes/includes to only output the one appropriate sln file.

**Reply by neman** [reply]
Thanks a lot!
